### DISCUSSION TITLE
How to open a new wezterm window in MacOS

### DISCUSSION THREAD

**Discussion by zhou13** [wezterm/wezterm#1208] [Q&A]
Hi all,

I want to create a global shortcut to open a new terminal. If I execute `wezterm` or `wezterm --start`, then I am able to open a new wezterm. However, there will be multiple icons in my taskbar.  I can also execute `wezterm cli spawn --new-window` which does not create new icons in the taskbar, but it is only effective when executed inside wezterm itself. I wonder whether there a way to open a new wezterm anywhere without spawning many icons in the taskbar in MacOS?

**Comment by azzamsa** [comment]
Same here.

Yes. In emacs dired I have a shortcut to open new tmux pane with current dired directory as argument.
Is is possible to open new pane/window in the **existing**/**running** wezterm using a shell script?

**Reply by wez** [reply]
Check out `wezterm cli spawn --help` and `wezterm cli split-pane --help`.  They locate the existing wezterm via the `WEZTERM_UNIX_SOCKET` environment that is passed down from wezterm to the processes spawned inside it, so if you're running emacs in wezterm, processes spawned by that emacs should inherit it and find the right wezterm.

Those don't work for the OP because they have a shortcut in their window environment and that doesn't know how to resolve the appropriate socket path to a running instance from outside of its process hierarchy.

**Reply by azzamsa** [reply]
I mean something like this: [open-external-terminal-and-tmux-from-dired](https://azzamsa.com/n/scripts-el/#open-external-terminal-and-tmux-from-dired).

```elisp
(defun term-here ()
  (interactive)
  (start-process "" nil "st"
                 "-e" "bash"
                 "-c" "tmux -q has-session && exec tmux new-window || exec tmux new-session -n$USER -s$USER@$HOSTNAME"))
```

![image](https://azzamsa.com/images/posts/emacs-scripts-7.webp)

 I tried to run `wezterm cli ...` from GUI Emacs, and it doesn't talk with existing Wezterm terminal.

**Reply by MuhammedZakir** [reply]
You can set `WEZTERM_UNIX_SOCKET` before running `wezterm cli spawn`. For example, in Linux:
```zsh
# this is where WezTerm puts its socket in my system.
for socket in /run/user/1000/wezterm/gui-sock-*
do
  [ -S "$socket" ] && break;
done

if [ -n "$socket" ]
then
  env WEZTERM_UNIX_SOCKET="$socket" wezterm cli spawn
else
  wezterm-gui start
fi
```

Edit: This works only if there is one WezTerm session running, or if you want spawn the tab in the oldest running session (which is what I want). If you want to spawn the tab in the newest session, remove the line inside `for` loop.

Edit-2: Fixed comment and added `else` branch to start WezTerm if it's not running.

**Reply by azzamsa** [reply]
Thanks a lot for the idea @MuhammedZakir. However, I found some issues.

My first implementation won't work if invoked outside wezterm. After the debug, I got:

``` bash
# invoked inside emacs

$ wezterm-here
$ wezterm-here
::: Socket found: /run/user/1000/wezterm/gui-sock-180639
 2021-10-14T20:56:50.122Z ERROR wezterm > --pane-id was not specified and $WEZTERM_PANE
                                    is not set in the environment; terminating
$
```

So I tried to add:

``` diff
if [ -n "$socket" ]
then
    echo "::: Socket found: $socket"
-    env WEZTERM_UNIX_SOCKET="$socket" wezterm cli spawn 
+   env WEZTERM_UNIX_SOCKET="$socket" wezterm cli spawn --pane-id 0
```

I have no clue which one is correct since using `--pane-id 0` and `--window-id 0` gives me the expected outcome. I want to create a new *window* at the end of other existing windows.

---

Now I have a fully working implementation, thanks!


```elisp
(defun term-here ()
  (interactive)
  (message "Opening terminal in current directory...")
  (start-process "" nil "wezterm-here" default-directory))
````

```bash
#!/bin/bash
# wezterm-here

cwd=$HOME

if [ -n "$1" ]; then
    cwd="$1"
fi

for socket in /run/user/1000/wezterm/gui-sock-*
do
  [ -S "$socket" ] && break;
done

if [ -n "$socket" ]
then
    echo "::: Socket found: $socket"
    env WEZTERM_UNIX_SOCKET="$socket" wezterm cli spawn --window-id 0 --cwd "$cwd"
fi
```

Now I can do `M-x term-here` anywhere in dired, and wezterm gives me a new window with the desired path.
Hope this helps any future googlers/duckduckgoers.

**Comment by rawsp33d** [comment]
Did you ever figure this out @zhou13? The answers given work for me in Linux but I haven't yet gotten anything to work in Macos. And I didn't even have this problem on Linux, as running `wezterm start` has been fine for me without any such problem.

**Reply by zhou13** [reply]
No, I cannot get it to work on MacOS. Therefore, I am still using alacritty.

**Reply by rawsp33d** [reply]
@wez I noticed something that might be of use. The native `cmd+n` shortcut actually opens a new window as expected. You can open as many new windows as you want---it doesn't stop at one for some reason.

So I'm wondering if whatever wezterm registers to place nicely with the OS here could be exposed or used by the `wezterm start` or `wezterm cli spawn --new-window` or something.

**Reply by wez** [reply]
For the past couple of releases, `wezterm start` automatically opens new windows using the already running wezterm gui process on all platforms

**Reply by yoplat** [reply]
> No, I cannot get it to work on MacOS. Therefore, I am still using alacritty.

Oh man after looking stuff up on how to do this, I've finally arrived at a solution that works for me (using skhd and yabai for window managment) by using applescript:
This is the script (openwez.scpt):
```
set appName to "WezTerm"

if application appName is running then
  Do Shell Script "/Applications/WezTerm.app/Contents/MacOS/wezterm-gui"
else
  tell application appName to activate
end if
```
And in skhd (keyboard daemon for launching commands with keybinds):
```
cmd - return: osascript ~/Scripts/openwez.scpt
```
If an wezterm is not running it opens it normally, but if it's already open it uses the weztern-gui command to open a new window. 
Other solutions didn't work for me for various reason:
- yabai (my window manager) didn't autofocus on the first window spawned by wezterm start for some reason
- using applescript and telling an already opened app to use Cmd+N to create a new instance would first focus on that app and then creating the window, this means that it would open the new window in the same space as the first one instead of the current one.
- cluttering the dock with multiple instances

This fixes all of the above and doesn't seem to impact performance.
Leaving it here for anyone looking for the solution.

**Comment by MelvinTo** [comment]
In macos, the unix sock file is in a different location, this works for me:

```bash
WEZTERM_UNIX_SOCKET=~/.local/share/wezterm/default-org.wezfurlong.wezterm wezterm cli spawn --new-window
```

**Comment by aur3l14no** [comment]
Why is wezterm start not behaving like wezterm cli spawn? Because spawning a window with "wezterm-gui" as title and with no bundleID seems a bit strange to me, I wonder if there are some considerations behind this? Thank you.
